File: cpp/tests/test_series_evo.cpp
// #define CATCH_CONFIG_MAIN
#define CATCH_CONFIG_RUNNER
#include <catch.hpp>

// Run tests in declaration order to match the Python test file.
int main(int argc, char* argv[]) {
	Catch::Session session;
	int result = session.applyCommandLine(argc, argv);
	if (result != 0) return result;
	session.configData().runOrder = Catch::RunTests::InDeclarationOrder;
	return session.run();
}

#include <iostream>
#include <iomanip>
#include <string>

#include <mpreal.h>
#include <symengine/expression.h>
#include <symengine/symbol.h>
#include <symengine/real_mpfr.h>

#include "expansions_evo_se.hpp"
#include "ellipse/ellipse.hpp"

using mpfr::mpreal;
using SymEngine::Expression;
using SymEngine::symbol;
using SymEngine::real_mpfr;
using SymEngine::mpfr_class;

static const mpfr_prec_t BITS    = 167;   // ~50 decimal places
static const int         MAX_ORD = 11;
static const mpreal      TOL("1e-9", BITS);

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

static mpreal ev(const Expression& expr, const SymEngine::map_basic_basic& s) {
	return mpreal(rcp_static_cast<const SymEngine::RealMPFR>(
			Expression(expr.get_basic()->subs(s)).get_basic()
	)->as_mpfr().get_mpfr_t());
}

static void assert_close(const std::string& title,
						 const mpreal& expected,
						 const mpreal& actual,
						 const mpreal& tol) {
	const mpreal abs_err = abs(actual - expected);
	const mpreal rel_err = (expected != 0) ? abs_err / abs(expected) : mpreal("inf");
	std::cout << "\n" << title << "\n"
			  << "  expected: " << expected << "\n"
			  << "  actual:   " << actual   << "\n"
			  << "  abs err:  " << abs_err  << "  (tol " << tol << ")\n"
			  << "  rel err:  " << rel_err  << "\n";
	REQUIRE(abs_err < tol);
}

// ---------------------------------------------------------------------------
// Fixture: reference values + substitution map for one polar test point
// ---------------------------------------------------------------------------

struct RefEvo {
	mpreal psi, rho, phi, h, sgn, abs_sin_psi, cos_psi;
	SymEngine::map_basic_basic subs;

	explicit RefEvo(const std::string& psi_deg) {
		set_precision_bits(BITS);
		std::cout << std::setprecision(50) << std::fixed;

		psi = mpreal(psi_deg) / mpreal("180.0") * mpfr::const_pi();
		rho = mpreal("5000.0");

		const mpreal x = rho * mpfr::cos(psi);
		const mpreal y = rho * mpfr::sin(psi);

		auto pair = mp_cartesian_to_ellipse(x, y);
		phi = pair.first;
		h   = pair.second;

		// psi is defined in (-180, 180] deg, so sign(psi) == sign(sin psi).
		sgn         = (psi > 0) ? mpreal(1) : mpreal(-1);
		abs_sin_psi = mpfr::abs(mpfr::sin(psi));
		// Signed cos(psi): its sign carries the x-sign, so the reconstructions
		// below are correct in every quadrant without an explicit reflection.
		cos_psi     = mpfr::cos(psi);

		const mpreal rho_ae2_val = rho / (mp_a() * mp_e2());
		const mpreal b_a_val     = mp_b() / mp_a();

		subs[symbol("sin_psi")] = real_mpfr(mpfr_class(abs_sin_psi.toString(), BITS));
		subs[symbol("rho_ae2")] = real_mpfr(mpfr_class(rho_ae2_val.toString(), BITS));
		subs[symbol("b_a")]     = real_mpfr(mpfr_class(b_a_val.toString(),     BITS));
	}
};

// ---------------------------------------------------------------------------
// Inside-evolute series, one polar test point per quadrant. psi (the polar
// angle) is defined in (-180, 180] deg (see RefEvo), which keeps the quadrant
// sign sgn = sign(sin psi) valid and exercises all four quadrants.
// ---------------------------------------------------------------------------

TEST_CASE("inside-evolute series (all quadrants)", "[series_evo]") {
	const std::string psi_deg = GENERATE(std::string("42.0"), std::string("138.0"),
										 std::string("-138.0"), std::string("-42.0"));
	const RefEvo r(psi_deg);
	INFO("psi = " << psi_deg << " deg");

	// (phi - sgn*pi/2) / (sgn * cos(psi)): the signed cos(psi) makes this
	// correct in every quadrant, so the series compares directly.
	assert_close("phi  evo sparse",
				 (r.phi - r.sgn * mpfr::const_pi() / 2) / (r.sgn * r.cos_psi),
				 ev(phi_evo_sparse(MAX_ORD, MAX_ORD), r.subs), TOL);
	assert_close("phi  evo dense",
				 (r.phi - r.sgn * mpfr::const_pi() / 2) / (r.sgn * r.cos_psi),
				 ev(phi_evo_dense(MAX_ORD), r.subs), TOL);

	// (sin(phi) - sgn) / sgn  (sin(phi) is invariant under the reflection).
	assert_close("(sin(phi)-sgn)/sgn  evo dense", (mpfr::sin(r.phi) - r.sgn) / r.sgn,
				 ev(sin_phi_evo_dense(MAX_ORD), r.subs), TOL);

	// cos(phi): series is cos(phi)/cos(psi); multiply by signed cos(psi).
	assert_close("cos(phi)  evo dense", mpfr::cos(r.phi),
				 r.cos_psi * ev(cos_phi_evo_dense(MAX_ORD), r.subs), TOL);

	// h in metres.
	assert_close("h [m]  evo sparse", r.h,
				 ev(h_evo_sparse(MAX_ORD, MAX_ORD), r.subs) * mp_a() + r.rho * r.abs_sin_psi,
				 TOL * mp_a());
	assert_close("h [m]  evo dense", r.h,
				 ev(h_evo_dense(MAX_ORD), r.subs) * mp_a() - mp_b() + r.rho * r.abs_sin_psi,
				 TOL * mp_a());
}

File: python/tests/test_series_evo.py
"""Pytest tests for the inside-evolute point-to-ellipse Fourier series.

Uses the WGS84 Earth ellipse with a fixed test point inside the evolute
(polar angle 138°, radius 5000 m). All assertions compare a truncated
series (order MAX_ORDER_EVO) against the multi-precision reference value.
"""

from mpmath import mp
mp.dps = 50

import pytest

from pytest_util import assert_close
from ellipse import mp_polar_to_cartesian, mp_cartesian_to_ellipse, mp_e2, mp_a
from symbols import sin_psi, rho_ae2, b_a
import fourier_series

# Order of the tested series.
MAX_ORDER_EVO = 11

# convergence parameter rho_ae2 = rho/(a*e2) ≈ 0.117, so rho_ae2^12 ≈ 5e-12.
TOL_EVO = mp.mpf("1e-9")

# Test point: polar coords inside the ellipse evolute.
_PSI_EVO = mp.mpf("138.") / mp.mpf("180.") * mp.pi
_RHO_EVO = mp.mpf("5000.")


@pytest.fixture(scope="module")
def ref_evo():
    cx, cy = mp_polar_to_cartesian(_PSI_EVO, _RHO_EVO)
    phi, h = mp_cartesian_to_ellipse(cx, cy)
    sgn = mp.mpf("1") if _PSI_EVO > 0 else mp.mpf("-1")
    return dict(
        psi         = _PSI_EVO,
        rho         = _RHO_EVO,
        phi         = phi,
        h           = h,
        sgn         = sgn,
        abs_sin_psi = abs(mp.sin(_PSI_EVO)),
        abs_cos_psi = abs(mp.cos(_PSI_EVO)),
        rho_ae2_val = _RHO_EVO / (mp_a * mp_e2),
        b_a_val     = mp.sqrt(mp.mpf(1) - mp_e2),
    )


def ev_evo(expr, ref):
    """Substitute evo-series reference values into a symbolic expression.

    Uses abs(sin(psi)), rho/(a*e2), and b/a as the three substitution variables.
    Returns an mpmath mpf preserving full precision.
    """
    result = (expr
              .subs(sin_psi, ref["abs_sin_psi"])
              .subs(rho_ae2, ref["rho_ae2_val"])
              .subs(b_a,     ref["b_a_val"]))
    return mp.mpf(result._mpf_)


# ---------------------------------------------------------------------------
# (phi - sgn*pi/2) / (sgn * |cos(psi)|)
# ---------------------------------------------------------------------------

def test_phi_evo_dense(ref_evo):
    expected = ((ref_evo["phi"] - ref_evo["sgn"] * mp.pi / 2)
                / (ref_evo["sgn"] * ref_evo["abs_cos_psi"]))
    result = ev_evo(fourier_series.phi_evo_sin_pow_dense(MAX_ORDER_EVO, MAX_ORDER_EVO), ref_evo)
    assert_close("(phi - sgn*pi/2)/(sgn*|cos(psi)|)  evo_dense", expected, result, TOL_EVO)

def test_phi_evo_dense_m(ref_evo):
    expected = ((ref_evo["phi"] - ref_evo["sgn"] * mp.pi / 2)
                / (ref_evo["sgn"] * ref_evo["abs_cos_psi"]))
    result = ev_evo(fourier_series.phi_evo_sin_pow_dense_m(MAX_ORDER_EVO), ref_evo)
    assert_close("(phi - sgn*pi/2)/(sgn*|cos(psi)|)  evo_dense", expected, result, TOL_EVO)


# ---------------------------------------------------------------------------
# (sin(phi) - sgn) / sgn
# ---------------------------------------------------------------------------

def test_sin_phi_evo_dense(ref_evo):
    expected = (mp.sin(ref_evo["phi"]) - ref_evo["sgn"]) / ref_evo["sgn"]
    result = ev_evo(fourier_series.sin_phi_evo_dense(MAX_ORDER_EVO, MAX_ORDER_EVO), ref_evo)
    assert_close("(sin(phi) - sgn)/sgn  evo_dense", expected, result, TOL_EVO)


# ---------------------------------------------------------------------------
# cos(phi) / |cos(psi)|
# ---------------------------------------------------------------------------

def test_cos_phi_evo_dense(ref_evo):
    expected = mp.cos(ref_evo["phi"]) / ref_evo["abs_cos_psi"]
    result = ev_evo(fourier_series.cos_phi_evo_dense(MAX_ORDER_EVO, MAX_ORDER_EVO), ref_evo)
    assert_close("cos(phi)/|cos(psi)|  evo_dense", expected, result, TOL_EVO)


# ---------------------------------------------------------------------------
# h in metres
# ---------------------------------------------------------------------------

def test_h_evo(ref_evo):
    result = ev_evo(fourier_series.h_a_evo(MAX_ORDER_EVO, MAX_ORDER_EVO), ref_evo) * mp_a
    assert_close("h [m]  evo", ref_evo["h"], result, TOL_EVO * mp_a)


def test_h_evo_dense(ref_evo):
    result = ev_evo(fourier_series.h_a_evo_dense(MAX_ORDER_EVO, MAX_ORDER_EVO), ref_evo) * mp_a
    assert_close("h [m]  evo_dense", ref_evo["h"], result, TOL_EVO * mp_a)